packages/orm/src/types.ts

```diff
@@ -11,6 +11,155 @@ import {
+  ManyToMany = "manyToMany"
+}
+
+/**
+ * Returns the type of the passed model class.
+ */
+export type ModelClassType<M extends AnyModel> = ReturnType<M["getClass"]>;
+
+/**
+ * Imitates the model bound to the session.
+ * 
+ * ModelId type helps to ensure that every model's instance has the id included within its type declaration.
+ */
+export type SessionBoundModel<
+  MClass extends AnyModel = AnyModel,
+> = MClass & { id?: ModelId };
+
+/**
+ * Infers the fields type passed to a model. 
+ */
+export type ModelFields<MClass extends AnyModel> = ConstructorParameters<
+  ModelClassType<MClass>
+> extends [infer FirstConstructorParam]
+ ? FirstConstructorParam extends ModelRefLike<infer ModelFields>
+   ? ModelFields
+   : never 
+ : never;
+
+ 
+ export type ModelConstructor<MClass extends AnyModel = AnyModel> = {
+   new (props: Ref<MClass>): SessionBoundModel<MClass>; 
+  }
+  export type QuerySetConstructor<MClassType extends typeof AnyModel, Payload extends object = {}> = {
+  new (modelClass: typeof AnyModel, clauses?: QueryClause<Payload>[], opts?: object): QuerySet<MClassType>;
+}
+
+/**
+ * Handles relationships on the source model side.
+ *
+ * Each key points either to a single entity or a query set of entities on another model.
+ */
+export type TargetRelationship<
+  MClass extends AnyModel,
+  Relation extends Relations
+> = Relation extends Relations.OneToOne
+  ? SessionBoundModel<MClass>
+  : Relation extends Relations.ForeignKey
+    ? SessionBoundModel<MClass>
+    : Relation extends Relations.ManyToMany
+      ? QuerySet<ModelClassType<MClass>>
+      : never;
+      
+/**
+ * Handles relationships on the target model side.
+ *
+ * Defines keys used to access the Model foreign key is being defined from, from the target model.
+ */
+export type SourceRelationship<
+  MClassType extends typeof AnyModel,
+  Relation extends Relations
+> = Relation extends Relations.OneToOne
+  ? InstanceType<MClassType>
+  : Relation extends Relations.ForeignKey
+    ? QuerySet<MClassType>
+    : Relation extends Relations.ManyToMany
+      ? QuerySet<MClassType>
+      : never;
+      
+/**
+ * Possible types of relations and attributes that describe a single model.
+ */
+type ModelField = QuerySet<any> | AnyModel | Serializable;
+  
+/**
+ * A map of possible types of relations and attributes that describe a single model.
+ */
+export type ModelFieldMap = {
+  id?: ModelId;
+  [K: string]: ModelField;
+};
+  
+/**
+ * A plain JS object representing the database entry.
+ */
+export type Ref<MClass extends AnyModel> = ConstructorParameters<
+  ModelClassType<MClass>
+> extends [infer FirstConstructorParam]
+  ? FirstConstructorParam extends ModelRefLike
+  ? FirstConstructorParam
+  : never
+  : never;
+
+/**
+ * Extends the plan JS object interface, representing the database entry, with {@link ModelField}
+ * 
+ * Mainly used in functions that manipulate db entries.
+ */
+export type RefWithFields<M extends AnyModel> = {
```

Comment:
This type is needed as we can be making db entries manipulations using not only serializable data but the model instances / nullish values as well.